### PR TITLE
 add headless-browser. Add condition when +renderer+ isn't set 

### PR DESCRIPTION
**Pull request by K1D77A**
Added headless-browser class. Add condition when +renderer+ isn't set. Also cleaned up the options so no more (getf ..) 
 I dont know if the commits are correct this is the first time I've contributed to a project like this sorry :(

### CODE REVIEW

File: source/browser.lisp [hunk 1]
```diff
@@ -238,8 +238,13 @@ windows.
 A typical Nyxt session encompasses a single instance of this class, but nothing
 prevents otherwise."))
 
+(define-class headless-browser (browser)
```
Comment (aartaka) [inline source/browser.lisp:241]: Huh, you seem to like this solution :D

However, do we really need it? Is there another way things can be done in, without introducing yet another fundamental entity?

File: source/start.lisp [hunk 2]
```diff
@@ -465,9 +465,11 @@ Examples:
   (nyxt:start :urls '(\"https://nyxt.atlas.engineer\" \"https://en.wikipedia.org\")
               :verbose t
               :with-file '(\"history\" \"/tmp/nyxt/history.lisp\"))"
-          (with-output-to-string (s) (opts:describe :stream s)))
+            (with-output-to-string (s) (opts:describe :stream s)))
   ;; Extensions should be made accessible straight from the beginning,
   ;; e.g. before a script is run.
+  (unless +renderer+
```
Comment (aartaka) [inline source/start.lisp:471]: I'm thinking of whether the error is the right approach. Maybe rather do `log:warn`? While I cannot imagine one loading bare Nyxt and starting it, there could be use-cases for it...

File: source/start.lisp [hunk 3]
```diff
@@ -476,54 +478,55 @@ Examples:
 
   ;; Options should be accessible anytime, even when run from the REPL.
   (setf *options* options)
-
```
Comment (aartaka) [inline source/start.lisp:479]: Ha, that's a smart one! This code is there for a really long time, and no one thought that destructuring can clarify it.

@Ambrevar, any reason it was done with `getf`? I remember that dome operations on plists require understanding whther the key is set to `nil` or missing, but that's not the case, right?

File: source/start.lisp [hunk 4]
```diff
@@ -571,10 +574,13 @@ Finally, run the browser, load URL-STRINGS if any, then run
                  (echo-warning "~a." message)
                  (error-in-new-window "*Init file errors*" full-message)))))
       (load-or-eval :remote nil)
-      (setf *browser* (make-instance 'user-browser
-                                     :startup-error-reporter-function startup-error-reporter
-                                     :startup-timestamp startup-timestamp
-                                     :socket-thread thread))
+      (setf *browser* (make-instance
+                       (if *headless-p*
+                           'headless-user-browser
```
Comment (aartaka) [inline source/start.lisp:579]: It probably should be `user-headless-browser`. 

Anyway, I'm still hesitant about accepting the necessity of `headless-browser`...

File: source/start.lisp (hunk 2 above)
Comment (K1D77A) [reply source/start.lisp:471]: Okay sounds like a plan :+1:

File: source/browser.lisp (hunk 1 above)
Comment (K1D77A) [reply source/browser.lisp:241]: Yes I am a big fan of CLOS. Seems to me that using a subclass of the normal class will allow the user to create specific behaviour like I had done previously.

Comment (aartaka) [reply source/browser.lisp:241]: Yes, it's convenient, but the same thing can be done with an `:around` method (also part of CLOS) and a check for `*headless-p*`. Introduction of another entity is not strictly necessary here.

**Comment by aartaka** [comment]
You're gonna hate me after this request, but can you squash these commits into one now that the changes are more or less homogenous?

Indentation and addition of blank lines (are they necessary?) belong to a separate commit (or even PR), as an exception to the rule.

**Comment by K1D77A** [comment]
> You're gonna hate me after this request, but can you squash these commits into one now that the changes are more or less homogenous?
> 
> Indentation and addition of blank lines (are they necessary?) belong to a separate commit (or even PR), as an exception to the rule.

I keep trying to squash the commits but git complains. I will just make a new one tomorrow :facepalm:

**Comment by aartaka** [comment]
No problem, take your time :)

File: source/start.lisp (hunk 3 above)
Comment (Ambrevar) [reply source/start.lisp:479]: No reason I believe.

**Pull request closed without merging**